### PR TITLE
Abstract admin actions and make them filterable for Discount Codes, Membership Levels, and Orders

### PR DESCRIPTION
**Pull request by mircobabini**
### All Submissions:

* [x] Have you followed the [Contributing guideline](https://github.com/strangerstudios/paid-memberships-pro/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/strangerstudios/paid-memberships-pro/pulls) for the same update/change?

### Changes proposed in this Pull Request:

1. Enhance code readability of extra row actions for orders
2. Add a filter for extra row actions on discountcodes and membershiplevels

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

### CODE REVIEW

**Comment by ideadude** [comment]
Should we put the built in actions into the array that gets filtered? That way, folks can use the same filter to remove, edit, or rearrange items for some reason?

FWIW, here's how we do actions in the members list table. Note the code to figure out the separated based on the spot in the list. https://github.com/strangerstudios/paid-memberships-pro/blob/dev/classes/class-pmpro-members-list-table.php#L443-L456

Here's another example where we do actions on the frontend: https://github.com/strangerstudios/paid-memberships-pro/blob/4812f157e759a38b2d22cf56be728cc00592af07/shortcodes/pmpro_account.php#L175-L198

**Comment by sc0ttkclark** [comment]
That sounds like a good idea @ideadude

**Comment by mircobabini** [comment]
I’ll check these out. I think the best thing we can do is to add something that we can support with an easy backward compatilibity when we will switch to the native WP_List_Table.

**Comment by sc0ttkclark** [comment]
Sounds like the array filter will maximize our compatibility when switching to the WP_List_Table so that sounds like a great direction with that as a goal

File: adminpages/orders.php [hunk 1]
```diff
@@ -1382,17 +1382,21 @@ class="alternate"<?php } ?>>
...
-									++ $i;
-									( $i == $action_count ) ? $sep = '' : $sep = ' | ';
-									$out .= "<span class='" . esc_attr( $action ) . "'>" . $link . $sep . "</span>";
-								}
-								echo $out;
+							/**
+							 * Filter the extra actions for this user on this order
+							 *
+							 * @param stdClass $order->user The user data
+							 * @param MemberOrder $order The current order
```
Comment (sc0ttkclark) [inline adminpages/orders.php:1389]:
```suggestion
							 * @param array.      $actions The list of actions.
							 * @param stdClass    $user    The user data.
							 * @param MemberOrder $order   The current order.
```

File: adminpages/membershiplevels.php [hunk 2]
```diff
@@ -835,6 +835,27 @@ function update_level_order(event, ui) {
...
 						<span class="copy"><a title="<?php _e('Copy', 'paid-memberships-pro' ); ?>" href="<?php echo add_query_arg( array( 'page' => 'pmpro-membershiplevels', 'edit' => -1, 'copy' => $level->id ), admin_url( 'admin.php' ) ); ?>"><?php _e('Copy', 'paid-memberships-pro' ); ?></a></span> |
 						<span class="delete"><a title="<?php _e('Delete', 'paid-memberships-pro' ); ?>" href="javascript:pmpro_askfirst('<?php echo str_replace("'", "\'", sprintf(__("Are you sure you want to delete membership level %s? All subscriptions will be cancelled.", 'paid-memberships-pro' ), $level->name));?>', '<?php echo wp_nonce_url(add_query_arg( array( 'page' => 'pmpro-membershiplevels', 'action' => 'delete_membership_level', 'deleteid' => $level->id ), admin_url( 'admin.php' ) ), 'delete_membership_level', 'pmpro_membershiplevels_nonce'); ?>'); void(0);"><?php _e('Delete', 'paid-memberships-pro' ); ?></a></span>
+
+						<?php
+						// Set up the hover actions for this level
```
Comment (sc0ttkclark) [inline adminpages/membershiplevels.php:840]:
```suggestion
```

File: adminpages/membershiplevels.php [hunk 3]
```diff
@@ -835,6 +835,27 @@ function update_level_order(event, ui) {
...
+						/**
+						 * Filter the extra actions for this level
+						 *
+						 * @since TBD
+						 *
+						 * @param stdClass $level The membership level data
```
Comment (sc0ttkclark) [inline adminpages/membershiplevels.php:846]:
```suggestion
						 * @param array  $actions The list of actions.
						 * @param object $level   The membership level data.
```

File: adminpages/membershiplevels.php [hunk 4]
```diff
@@ -835,6 +835,27 @@ function update_level_order(event, ui) {
...
 						<span class="copy"><a title="<?php _e('Copy', 'paid-memberships-pro' ); ?>" href="<?php echo add_query_arg( array( 'page' => 'pmpro-membershiplevels', 'edit' => -1, 'copy' => $level->id ), admin_url( 'admin.php' ) ); ?>"><?php _e('Copy', 'paid-memberships-pro' ); ?></a></span> |
 						<span class="delete"><a title="<?php _e('Delete', 'paid-memberships-pro' ); ?>" href="javascript:pmpro_askfirst('<?php echo str_replace("'", "\'", sprintf(__("Are you sure you want to delete membership level %s? All subscriptions will be cancelled.", 'paid-memberships-pro' ), $level->name));?>', '<?php echo wp_nonce_url(add_query_arg( array( 'page' => 'pmpro-membershiplevels', 'action' => 'delete_membership_level', 'deleteid' => $level->id ), admin_url( 'admin.php' ) ), 'delete_membership_level', 'pmpro_membershiplevels_nonce'); ?>'); void(0);"><?php _e('Delete', 'paid-memberships-pro' ); ?></a></span>
+
+						<?php
+						// Set up the hover actions for this level
+						/**
+						 * Filter the extra actions for this level
```
Comment (sc0ttkclark) [inline adminpages/membershiplevels.php:842]:
```suggestion
						 * Filter the extra actions for this level.
```

File: adminpages/orders.php [hunk 5]
```diff
@@ -1382,17 +1382,21 @@ class="alternate"<?php } ?>>
...
-							if ( $action_count ) {
-								$out = ' | ';
-								foreach ( $actions as $action => $link ) {
-									++ $i;
-									( $i == $action_count ) ? $sep = '' : $sep = ' | ';
-									$out .= "<span class='" . esc_attr( $action ) . "'>" . $link . $sep . "</span>";
-								}
-								echo $out;
+							/**
+							 * Filter the extra actions for this user on this order
```
Comment (sc0ttkclark) [inline adminpages/orders.php:1386]:
```suggestion
							 * Filter the extra actions for this user on this order.
```

File: adminpages/discountcodes.php [hunk 6]
```diff
@@ -818,6 +818,27 @@
 											<a title="<?php _e(' View Orders', 'paid-memberships-pro' ); ?>" href="<?php echo add_query_arg( array( 'page' => 'pmpro-orders', 'discount-code' => $code->id, 'filter' => 'with-discount-code' ), admin_url('admin.php' ) ); ?>"><?php _e( 'Orders', 'paid-memberships-pro' ); ?></a>
 										</span>
 									<?php } ?>
+                                    
+                                    <?php
+                                    // Set up the hover actions for this discount code
```
Comment (sc0ttkclark) [inline adminpages/discountcodes.php:823]:
```suggestion
```

File: adminpages/discountcodes.php [hunk 7]
```diff
@@ -818,6 +818,27 @@
...
+                                    /**
+                                     * Filter the extra actions for this discount code
```
Comment (sc0ttkclark) [inline adminpages/discountcodes.php:825]:
```suggestion
                                     * Filter the extra actions for this discount code.
```

File: adminpages/discountcodes.php [hunk 8]
```diff
@@ -818,6 +818,27 @@
...
+                                     *
+                                     * @since TBD
+                                     *
+                                     * @param stdClass $code The discount code data
```
Comment (sc0ttkclark) [inline adminpages/discountcodes.php:829]:
```suggestion
                                     * @param array  $actions The list of actions.
                                     * @param object $code    The discount code data.
```

**Review by sc0ttkclark — APPROVED** [review]

File: adminpages/membershiplevels.php [hunk 9]
```diff
@@ -832,9 +832,90 @@ function update_level_order(event, ui) {
...
+
+						/**
+						 * Filter the extra actions for this level.
+						 *
+						 * @since TBD
+						 *
+						 * @param array  $actions The list of actions.
+						 * @param object $level   The membership level data.
+						 */
+						$actions = apply_filters( 'pmpro_membershiplevels_row_actions', $actions, $code );
```
Comment (sc0ttkclark) [inline adminpages/membershiplevels.php:903]:
```suggestion
						$actions = apply_filters( 'pmpro_membershiplevels_row_actions', $actions, $level );
```

**Review by sc0ttkclark — APPROVED** [review]

**Comment by sc0ttkclark** [comment]
I made all of the required changes to bring the existing actions into the filterable action array for Discount Codes, Membership Levels, and Orders. I tested each on my side and all of the actions work as expected.

**Review by andrewlimaza — APPROVED** [review]

**Pull request merged**